### ISSUE TITLE
permessage-deflate compression support

### ISSUE THREAD

**Issue by gordol** [miguelgrinberg/python-socketio#207]
Official Socket.IO [since 1.4.0](https://socket.io/blog/socket-io-1-4-0/) supports compression via [permessage-deflate (rfc7692)](https://tools.ietf.org/html/rfc7692).

What would be needed to support that here? Or is it better to support that perhaps with a middleware or something?

**Comment by gordol** [comment]
FWIW, Eventlet's WebSocket server supports this since last year: https://github.com/eventlet/eventlet/pull/417

But looks like the eventlet implementation here isn't using their websocket class? Hmm...

**Comment by miguelgrinberg** [comment]
I'll have to see, because eventlet isn't the only websocket implementation I use. It's probably easy, and yes, I am using the WebSocket support included in eventlet, so if they have it I should be able to support it.

**Comment by gordol** [comment]
Hmm, does eventlet's normal wsgi server do some magic that makes it different than the websocket wsgi server?

I realize now I was looking at Flask-SocketIO code here: https://github.com/miguelgrinberg/Flask-SocketIO/blob/d8688bd0780641c18e585b7d8b3d938ce7d6a5e3/flask_socketio/__init__.py#L520

Looks to me like we're using eventlet's normal wsgi server, not the websocket server. The gevent implementation does use a websocket extension to the wsgi server, though, as shown here: https://github.com/miguelgrinberg/Flask-SocketIO/blob/d8688bd0780641c18e585b7d8b3d938ce7d6a5e3/flask_socketio/__init__.py#L539-L540).

I also dug into python-engineio, and you're right, it's using the WebSocketWSGI implementation: https://github.com/miguelgrinberg/python-engineio/blob/1a0243ca5d2bb67bdf9b261ad4df42ecd30b1742/engineio/async_eventlet.py

So, maybe this is actually already supported?

**Comment by miguelgrinberg** [comment]
Yeah, maybe it is all an internal enhancement and it just works if the client requests the compression.

**Comment by gordol** [comment]
I'm going to close this...

It took a bit of lsof/strace hackery, but I was finally able to trace the socket calls between gunicorn and nginx. I then reverted back to eventlet==0.21.0 but it thew an error `ImportError: Failed to initialize: no suitable implementation for this system`. So I tried eventlet==0.20.1, and was able to see the raw JSON data flowing by. I then switched back to 0.22.0 and I'm pretty sure the messages are compressed. ;)

Before:
![uncompressed](https://node.zeneval.com/snaps/0a33b02c1621f8e9345861f78180ee2b.png)

After:
![compressed](https://node.zeneval.com/snaps/7f800bc33e1ddad97afdfafc43561144.png)
